### DISCUSSION TITLE
Lesson 7: VirtualMachineError when calling endLottery function

### DISCUSSION THREAD

**Discussion by rsamban** [smartcontractkit/full-blockchain-solidity-course-py#627] [Q&A] [answered]
I am getting following error when calling endLottery **function.** Not sure what I am missing. Can you please help?

-----------

Brownie v1.17.2 - Python development framework for Ethereum

SmartcontractLotteryProject is the active project.

Launching 'ganache-cli --accounts 10 --hardfork istanbul --gasLimit 12000000 --mnemonic brownie --port 8545'...

Running 'scripts/deploy_lottery.py::main'...
Transaction sent: 0x13207324e29c4b959f1354eac94b6ed842b051dae3aad9f3db0c0c13ef4012a3
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 0
  MockV3Aggregator.constructor confirmed   Block: 1   Gas used: 430659 (3.59%)
  MockV3Aggregator deployed at: 0x3194cBDC3dbcd3E11a07892e7bA5c3394048Cc87

Transaction sent: 0xe32677f021ed8131944b28cbc57e3fe75a1b28b3f10041e89cebcfd90a8d338f
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 1
  LinkToken.constructor confirmed   Block: 2   Gas used: 669136 (5.58%)
  LinkToken deployed at: 0x602C71e4DAC47a042Ee7f46E0aee17F94A3bA0B6

Transaction sent: 0xc509c872ecebb1dddb54a37b841a38553731c2d925ee7c0034357eed4e32cb08
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 2
  VRFCoordinatorMock.constructor confirmed   Block: 3   Gas used: 276395 (2.30%)
  VRFCoordinatorMock deployed at: 0xE7eD6747FaC5360f88a2EFC03E00d25789F69291

Transaction sent: 0x344666c4ca24d48671f492c0f26abfac18fc0228e7d5e8a2002f3ec7289edcf8
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 3
  Lottery.constructor confirmed   Block: 4   Gas used: 896904 (7.47%)
  Lottery deployed at: 0x6951b5Bd815043E3F842c1b026b0Fa888Cc2DD85

network:::{'gas_limit': 'max', 'gas_buffer': 1, 'gas_price': 0, 'max_fee': None, 'priority_fee': None, 'reverting_tx_gas_limit': 'max', 'default_contract_owner': True, 'cmd_settings': None, 'keyhash': '0x2ed0feb3e7fd2022120aa84fab1945545a9f2ffc9076fd6156fa96eaff4c1311', 'fee': 100000000000000000}
Deployed Lottery
Transaction sent: 0x1f3629b4ccc88f7e9e8b4f9d7d7a0e3328610fafa9346bd79152467364540969
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 4
  Lottery.startLottery confirmed   Block: 5   Gas used: 28902 (0.24%)

  Lottery.startLottery confirmed   Block: 5   Gas used: 28902 (0.24%)

Lottery Started!
Transaction sent: 0xda214dddeb9bc33dab1c188730bf5ec64d0a996de7e367299176ac595944dfbf
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 5
  Lottery.enter confirmed   Block: 6   Gas used: 70995 (0.59%)

  Lottery.enter confirmed   Block: 6   Gas used: 70995 (0.59%)

Entered Lottery!
Transaction sent: 0xf7b3761cd940dde0b042ff692416df3b66be04dd22ececfdcd8f2230317a6cc7
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 6
  LinkToken.transfer confirmed   Block: 7   Gas used: 51386 (0.43%)

  LinkToken.transfer confirmed   Block: 7   Gas used: 51386 (0.43%)

Contract Finded!
  LinkToken.transfer confirmed   Block: 7   Gas used: 51386 (0.43%)

Transaction sent: 0xe97361a7db515a30e3feac79b1e14d6e16abd3132949b17a82fbb83cd87bc619
  Gas price: 0.0 gwei   Gas limit: 12000000   Nonce: 7
  Lottery.endLottery confirmed (reverted)   Block: 8   Gas used: 11813033 (98.44%)

  File "brownie/_cli/run.py", line 50, in main
    return_value, frame = run(
  File "brownie/project/scripts.py", line 103, in run
    return_value = f_locals[method_name](*args, **kwargs)
  File "./scripts/deploy_lottery.py", line 59, in main
    end_lottery()
  File "./scripts/deploy_lottery.py", line 48, in end_lottery
    tx = lottery.endLottery({"from": account})
  File "brownie/network/contract.py", line 1629, in __call__
    return self.transact(*args)
  File "brownie/network/contract.py", line 1502, in transact
    return tx["from"].transfer(
  File "brownie/network/account.py", line 682, in transfer
    receipt._raise_if_reverted(exc)
  File "brownie/network/transaction.py", line 446, in _raise_if_reverted
    raise exc._with_attr(
VirtualMachineError: revert
Trace step -1, program counter 2468:
  File "/Users/rsamban/.brownie/packages/smartcontractkit/chainlink-brownie-contracts@1.1.1/contracts/src/v0.6/VRFConsumerBase.sol", line 161, in VRFConsumerBase.requestRandomness:    
    function requestRandomness(bytes32 _keyHash, uint256 _fee)
      internal returns (bytes32 requestId)
    {
      LINK.transferAndCall(vrfCoordinator, _fee, abi.encode(_keyHash, USER_SEED_PLACEHOLDER));
      // This is the seed passed to VRFCoordinator. The oracle will mix this with
      // the hash of the block containing this request to obtain the seed/input
      // which is finally passed to the VRF cryptographic machinery.
Terminating local RPC client...

-----------------------

**Comment by cromewar** [comment, marked as answer]
Hello @rsamban as the error says the problem is with your request randomness function, could you please share here your contract code in order to check it?

**Reply by rsamban** [reply]
@cromewar  Thanks for your response. Following is how I can reproduce this.
$ git clone https://github.com/PatrickAlphaC/smartcontract-lottery.git smartcontract-lottery-orig
$ cd smartcontract-lottery-orig
$ brownie compile
$ brownie run scripts/deploy_lottery.py

// Above Works fine.

However, when I replace fund_with_link function in helpful_scripts.py with following I get the above error.
```Python
def fund_with_link(
    contract_address, account=None, link_token=None, amount=20000000000000000
):  # 0.1 LINK
    account = account if account else get_account()
    link_token = link_token if link_token else get_contract("link_token")
    tx = link_token.transfer(contract_address, amount, {"from": account})
    # link_token_contract = interface.LinkTokenInterface(link_token.address)
    # tx = link_token_contract.transfer(contract_address, amount, {"from": account})
    tx.wait(1)
    print("Contract Funded!")
    return tx
```

So I believe the problem is with this function, but I can't figure it out.
I can get my code working when I replace the one from the repo.
I can keep going now, but it is just annoying..., it is just one small comma or period, which is taking > couple of hours. Agh...

As requested, i have attached my Logger.sol also - just in case
[Lottery.TXT](https://github.com/smartcontractkit/full-blockchain-solidity-course-py/files/7776328/Lottery.TXT)
.

**Reply by cromewar** [reply]
Well the problem is with the amount as far as it seems, you are using `20000000000000000` which has 17 digits and you should use something with 18 digits, try this instead:

```Python
def fund_with_link(
    contract_address, account=None, link_token=None, amount=100000000000000000
):  # 0.1 link
    account = account if account else get_account()
    link_token = link_token if link_token else get_contract("link_token")
    tx = link_token.transfer(contract_address, amount, {"from": account})
    # link_token_contract = interface.LinkTokenInterface(link_token.address)
    # tx = link_token_contract.transfer(contract_address, amount, {"from": account})
    tx.wait(1)
    print("Fund contract successful")
    return tx

```

**Reply by rsamban** [reply]
Yes, it was, thanks @cromewar

**Reply by cromewar** [reply]
I'm glad is solved now.
